src/core/Grappa.sol

```diff
@@ -190,137 +192,99 @@ contract Grappa is OwnableUpgradeable, ReentrancyGuardUpgradeable, UUPSUpgradeab
-
+    function settle(address _account, uint256 _tokenId, uint256 _amount)
+        public
+        nonReentrant
+        returns (uint256 debt, uint256 payout)
+    {
         optionToken.burnGrappaOnly(_account, _tokenId, _amount);
 
-        IMarginEngine(engine).payCashValue(collateral, _account, payout);
+        Settlement memory settlement = _settle(_account, _tokenId, _amount, false);
 
-        return payout;
+        return (settlement.debt, settlement.payout);
     }
 
     /**
-     * @notice burn array of option tokens and get out cash value at expiry
+     * @notice burn array of option tokens and settles debts and payouts at expiry
      *
      * @param _account who to settle for
      * @param _tokenIds array of tokenIds to burn
-     * @param _amounts   array of amounts to burn
+     * @param _amounts array of amounts to burn
+     * @param _dryRun flag to simulate transaction
      */
-    function batchSettleOptions(address _account, uint256[] memory _tokenIds, uint256[] memory _amounts)
+    function batchSettle(address _account, uint256[] memory _tokenIds, uint256[] memory _amounts, bool _dryRun)
         external
         nonReentrant
-        returns (Balance[] memory payouts)
+        returns (Balance[] memory debts, Balance[] memory payouts)
     {
         if (_tokenIds.length != _amounts.length) revert GP_WrongArgumentLength();
 
-        if (_tokenIds.length == 0) return payouts;
-
-        optionToken.batchBurnGrappaOnly(_account, _tokenIds, _amounts);
+        if (_tokenIds.length == 0) return (debts, payouts);
 
-        address lastCollateral;
-        address lastEngine;
-
-        uint256 lastTotalPayout;
+        if (!_dryRun) optionToken.batchBurnGrappaOnly(_account, _tokenIds, _amounts);
 
         for (uint256 i; i < _tokenIds.length;) {
-            (address engine, address collateral, uint256 payout) = getPayout(_tokenIds[i], _amounts[i].toUint64());
-
-            uint8 collateralId = _tokenIds[i].parseCollateralId();
+            Settlement memory settlement = _settle(_account, _tokenIds[i], _amounts[i], _dryRun);
 
-            payouts = _addToPayouts(payouts, collateralId, payout);
+            if (settlement.debt != 0) {
+                debts = _addToBalances(debts, settlement.debtAssetId, settlement.debt);
+            }
 
-            // if engine or collateral changes, payout and clear temporary parameters
-            if (lastEngine == address(0)) {
-                lastEngine = engine;
-                lastCollateral = collateral;
-            } else if (engine != lastEngine || lastCollateral != collateral) {
-                IMarginEngine(lastEngine).payCashValue(lastCollateral, _account, lastTotalPayout);
-                lastTotalPayout = 0;
-                lastEngine = engine;
-                lastCollateral = collateral;
+            if (settlement.payout != 0) {
+                payouts = _addToBalances(payouts, settlement.payoutAssetId, settlement.payout);
```

Comment:
same as above, very necessary

Reply:
I'm hating the array appending. but yeah get that it might be the only way

Reply:
We could create an array with size equal to number of tokens and at the end shrink the array down. This would create only 1 additional  array.

Havent run the tests to see the gas implications, but could be worth a look.